New rule suggestion: literal conventions mixed

With `space-in-brackets` enabled I had an error like this:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.'].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

I got noticed by `space-in-brackets`

```
error  There should be no space after '['  space-in-brackets
```

Since there are certainly spaces after the opening bracket this rule fails correctly, but it's not the most clear message possible.

The same occurs when array is written in consistently, but the object literal is not:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.',
].join('\n'), {
    foo: 'foo',
    bar: 'bar'});
```

```
error  There should be no space after '{'  space-in-brackets
```
However using new lines inside the literal without spaces after left bracket/before right brakcet is ok for this rule:
 
```
_.template(['Foo: ${ foo }',
    'Bar: `${ bar }`.'].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

The error which caused the problem is the mix of conventions: first element of the array is in the new line, but there is no new line after the last. The correct form of the snippet (with both literals expanded) will be:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.',
].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

I suggest adding new rule rule which would warn if conventions are mixed.
Already existing rules do not seem to cover the case: I can either force or disable spaces in brackets but there is no rule which explicitly addresses the case checking whether the literal (both array and object i suppose) is written consistently.

My eslint configuration related to stylistic issues:
```
"brace-style": 2,
"camelcase": 0,
"new-cap": 0,
"no-lonely-if": 2,
"no-underscore-dangle": 0,
"space-after-keywords": 2,
"space-in-brackets": [2, "never"],
"space-infix-ops": 2,
"quotes": [2, "single"],
```

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/7409003-new-rule-suggestion-array-literal-conventions-mixed?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

I use eslint via grunt-eslint `2.1.0` and it has `^0.10.0` with `0.10.2` currently installed in my setup.
I checked on v0.11.0 as well without grunt and linting the following snippet:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.'].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

results in the same error:

```
  1:11  error  There should be no space after '['  space-in-brackets
```

--- a/docs/rules/space-in-brackets.md
+++ b/docs/rules/space-in-brackets.md
@@ -14,7 +14,7 @@ foo['bar'];
 
 ## Rule Details
 
-This rule aims to maintain consistency around the spacing inside of square brackets, either by disallowing spaces inside of brackets between the brackets and other tokens or enforcing spaces. Multi-line array and object literals with no values on the same line as the brackets are excepted from this rule as this is a common pattern.  Object literals that are used as the first or last element in an array are also ignored.
+This rule aims to maintain consistency around the spacing inside of square brackets, either by disallowing spaces inside of brackets between the brackets and other tokens or enforcing spaces. Brackets that are separated from the adjacent value by a new line are excepted from this rule, as this is a common pattern.  Object literals that are used as the first or last element in an array are also ignored.
 
 ### Options
 
@@ -36,9 +36,6 @@ When `"never"` is set, the following patterns are considered warnings:
 ```js
 foo[ 'bar' ];
 foo['bar' ];
-foo[
-    'bar'
-];
 
 var arr = [ 'foo', 'bar' ];
 var arr = ['foo', 'bar' ];
@@ -48,19 +45,10 @@ var arr = ['foo',
   'bar'
 ];
 
-var arr = [
-  'foo',
-  'bar'];
-
 var obj = { 'foo': 'bar' };
 var obj = {'foo': 'bar' };
 var obj = { baz: {'foo': 'qux'}, 'bar'};
 var obj = {baz: { 'foo': 'qux' }, 'bar'};
-var obj = {'foo': 'bar'
-};
-
-var obj = {
-  'foo':'bar'};
 ```
 
 The following patterns are not warnings:
@@ -69,6 +57,11 @@ The following patterns are not warnings:
 // When options are [2, "never"]
 
 foo['bar'];
+foo[
+  'bar'
+];
+foo[
+  'bar'];
 
 var arr = [];
 var arr = ['foo', 'bar', 'baz'];
@@ -79,6 +72,10 @@ var arr = [
   'baz'
 ];
 
+var arr = [
+  'foo',
+  'bar'];
+
 var obj = {'foo': 'bar'};
 
 var obj = {'foo': {'bar': 'baz'}, 'qux': 'quxx'};
@@ -86,6 +83,10 @@ var obj = {'foo': {'bar': 'baz'}, 'qux': 'quxx'};
 var obj = {
   'foo': 'bar'
 };
+var obj = {'foo': 'bar'
+};
+var obj = {
+  'foo':'bar'};
 
 var obj = {};
 ```

--- a/lib/rules/space-in-brackets.js
+++ b/lib/rules/space-in-brackets.js
@@ -124,18 +124,24 @@ module.exports = function(context) {
             var propertyNameMustBeSpaced = options.propertyNameException ?
                                     !options.spaced : options.spaced;
 
-            if (isSameLine(before, first) || isSameLine(last, after)) {
+            if (isSameLine(before, first)) {
                 if (propertyNameMustBeSpaced) {
                     if (!isSpaced(before, first) && isSameLine(before, first)) {
                         reportRequiredBeginningSpace(node, before);
-                    }
-                    if (!isSpaced(last, after) && isSameLine(last, after)) {
-                        reportRequiredEndingSpace(node, after);
                     }
                 } else {
                     if (isSpaced(before, first)) {
                         reportNoBeginningSpace(node, before);
                     }
+                }
+            }
+
+            if (isSameLine(last, after)) {
+                if (propertyNameMustBeSpaced) {
+                    if (!isSpaced(last, after) && isSameLine(last, after)) {
+                        reportRequiredEndingSpace(node, after);
+                    }
+                } else {
                     if (isSpaced(last, after)) {
                         reportNoEndingSpace(node, after);
                     }
@@ -165,13 +171,16 @@ module.exports = function(context) {
                 options.singleElementException && node.elements.length === 1
                 ? !options.spaced : options.spaced;
 
-            if (isSameLine(first, second) || isSameLine(penultimate, last)) {
+            if (isSameLine(first, second)) {
                 if (openingBracketMustBeSpaced && !isSpaced(first, second)) {
                     reportRequiredBeginningSpace(node, first);
                 }
                 if (!openingBracketMustBeSpaced && isSpaced(first, second)) {
                     reportNoBeginningSpace(node, first);
                 }
+            }
+
+            if (isSameLine(penultimate, last)) {
                 if (closingBracketMustBeSpaced && !isSpaced(penultimate, last)) {
                     reportRequiredEndingSpace(node, last);
                 }
@@ -196,13 +205,16 @@ module.exports = function(context) {
                 options.objectsInObjectsException && penultimate.value === "}"
                 ? !options.spaced : options.spaced;
 
-            if (isSameLine(first, second) || isSameLine(penultimate, last)) {
+            if (isSameLine(first, second)) {
                 if (options.spaced && !isSpaced(first, second)) {
                     reportRequiredBeginningSpace(node, first);
                 }
                 if (!options.spaced && isSpaced(first, second)) {
                     reportNoBeginningSpace(node, first);
                 }
+            }
+
+            if (isSameLine(penultimate, last)) {
                 if (closingCurlyBraceMustBeSpaced && !isSpaced(penultimate, last)) {
                     reportRequiredEndingSpace(node, last);
                 }

--- a/tests/lib/rules/space-in-brackets.js
+++ b/tests/lib/rules/space-in-brackets.js
@@ -97,6 +97,14 @@ eslintTester.addRuleTest("lib/rules/space-in-brackets", {
         { code: "obj[ obj2[ foo ] ]", args: [2, "never", {"propertyName": true}] },
         { code: "obj['for' + 'Each'](function (item) { return [\n1,\n2,\n3,\n4\n]; })", args: [2, "never"] },
 
+
+        { code: "obj[\nfoo]", args: [2, "never"] },
+        { code: "obj[foo\n]", args: [2, "never"] },
+        { code: "var obj = {foo: bar,\nbaz: qux\n};", args: [2, "never"] },
+        { code: "var obj = {\nfoo: bar,\nbaz: qux};", args: [2, "never"] },
+        { code: "var arr = [1,\n2,\n3,\n4\n];", args: [2, "never"] },
+        { code: "var arr = [\n1,\n2,\n3,\n4];", args: [2, "never"] },
+
         // never - singleValue
         { code: "var foo = [ 'foo' ]", args: [2, "never", {singleValue: true}] },
         { code: "var foo = [ 2 ]", args: [2, "never", {singleValue: true}] },
@@ -556,26 +564,6 @@ eslintTester.addRuleTest("lib/rules/space-in-brackets", {
             ]
         },
         {
-            code: "var obj = {foo: bar,\nbaz: qux\n};",
-            args: [2, "never"],
-            errors: [
-                {
-                    message: "There should be no space before '}'",
-                    type: "ObjectExpression"
-                }
-            ]
-        },
-        {
-            code: "var obj = {\nfoo: bar,\nbaz: qux};",
-            args: [2, "never"],
-            errors: [
-                {
-                    message: "There should be no space after '{'",
-                    type: "ObjectExpression"
-                }
-            ]
-        },
-        {
             code: "var arr = [1, 2, 3, 4];",
             args: [2, "always"],
             errors: [
@@ -667,26 +655,6 @@ eslintTester.addRuleTest("lib/rules/space-in-brackets", {
                 },
                 {
                     message: "There should be no space before ']'",
-                    type: "ArrayExpression"
-                }
-            ]
-        },
-        {
-            code: "var arr = [1,\n2,\n3,\n4\n];",
-            args: [2, "never"],
-            errors: [
-                {
-                    message: "There should be no space before ']'",
-                    type: "ArrayExpression"
-                }
-            ]
-        },
-        {
-            code: "var arr = [\n1,\n2,\n3,\n4];",
-            args: [2, "never"],
-            errors: [
-                {
-                    message: "There should be no space after '['",
                     type: "ArrayExpression"
                 }
             ]

